dcl_slam: {
  # frames
  world_frame: "world",                   # global frame
  odom_frame: "odom",                     # local frame

  # lidar configuration
  sensor: velodyne,                       # lidar type, support 'velodyne 16/64' or 'livox 6'
  n_scan: 16,                             # number of lidar channel (i.e., 6, 16, 64)

  # CPU params
  onboard_cpu_cores_num: 4,               # cores number of onboard unit
  loop_closure_process_interval: 0.01,    # interval of detecting loop (in second)
  map_publish_interval: 10.0,             # interval of publish global maps (in second)
  global_map_visualization_radius: 50,    # radius of global map (in meter)
  mapping_process_interval: 0.1,          # interval of optmization (in second)

  # keyframe params
  keyframe_distance_threshold: 1.0,       # keyframe distance threshold (in meter)
  keyframe_angle_threshold: 0.2,          # keyframe angle threshold (in rad)

  # mapping
  global_optmization_enable: true,        # enable distributed DGS
  use_between_noise: true,                # use between noise flag
  optmization_maximum_iteration: 30,      # maximum iterations time of optimization
  use_pcm: true,                          # enable pairwise consistency maximization (with descriptor_distance_threshold=0.5)
  pcm_threshold: 0.5,                     # confidence probability for PCM (i.e., 0.01, 0.05, 0.1, 0.25, 0.5, 0.75)

  # downsample
  descript_leaf_size: 0.1,                # descriptor downsample rate (default 0.1)
  map_leaf_size: 0.4,                     # scan to map matching downsample rate (default 0.4)

  # loop closure
  intra_robot_loop_closure_enable: true, # enable to search intra-robot loop closre with global descriptor
  inter_robot_loop_closure_enable: true,  # enable to search intra-robot loop closre with global descriptor
  descriptor_type: "LidarIris",           # descriptor type: ScanContext, LidarIris, M2DP
  knn_candidates: 6,                      # k nearest neighbor search of row key
  match_mode: 2,                          # iris-feature matching mode, (i.e., 0, 1, 2; default 2) 
  descriptor_distance_threshold: 0.4,     # iris-feature matching threshold
  exclude_recent_frame_num: 50,           # exclude recent keyframe in intra-robot loop closure
  search_radius: 30.0,                    # radius of radius search based intra-robot loop closure
  history_keyframe_search_num: 16,        # number of history frames in submap for scan-to-map matching
  fitness_score_threshold: 0.2,           # ICP fitness score threshold
  ransac_maximum_iteration: 2000,         # RANSAC maximum iteration time
  ransac_threshold: 0.3,                  # RANSAC threshold (rate: [0 1])
  ransac_outlier_reject_threshold: 0.35,  # RANSAC outlier rejection distancce

  # output
  save_directory: "/dcl_output",          # save at $HOME/saveDistributedTrajectoryDirectory
}

lio_sam:

  # Topics
  pointCloudTopic: "velodyne_points"          # Point cloud data
  imuTopic: "imu/data"                        # IMU data
  odomTopic: "odometry/imu"                   # IMU pre-preintegration odometry, same frequency as IMU
  gpsTopic: "odometry/gpsz"                   # GPS odometry topic from navsat, see module_navsat.launch file

  # Frames
  lidarFrame: "base_link"
  baselinkFrame: "base_link"
  odometryFrame: "odom"
  mapFrame: "map"

  # GPS Settings
  useImuHeadingInitialization: false           # if using GPS data, set to "true"
  useGpsElevation: false                      # if GPS elevation is bad, set to "false"
  gpsCovThreshold: 2.0                        # m^2, threshold for using GPS data
  poseCovThreshold: 25.0                      # m^2, threshold for using GPS data
  
  # Export settings
  savePCD: false                              # https://github.com/TixiaoShan/LIO-SAM/issues/3
  savePCDDirectory: "/Downloads/LOAM/"        # in your home folder, starts and ends with "/". Warning: the code deletes "LOAM" folder then recreates it. See "mapOptimization" for implementation

  # Sensor Settings
  sensor: velodyne                            # lidar sensor type, 'velodyne' or 'ouster' or 'livox'
  N_SCAN: 16                                  # number of lidar channel (i.e., Velodyne/Ouster: 16, 32, 64, 128, Livox Horizon: 6)
  Horizon_SCAN: 1800                          # lidar horizontal resolution (Velodyne:1800, Ouster:512,1024,2048, Livox Horizon: 4000)
  downsampleRate: 1                           # default: 1. Downsample your data if too many points. i.e., 16 = 64 / 4, 16 = 16 / 1
  lidarMinRange: 1.0                          # default: 1.0, minimum lidar range to be used
  lidarMaxRange: 1000.0                       # default: 1000.0, maximum lidar range to be used

  # IMU Settings
  imuAccNoise: 1.0557715622632996e-02
  imuGyrNoise: 1.0084886221461694e-02
  imuAccBiasN: 2.3084619195565281e-04
  imuGyrBiasN: 9.8778648600726277e-05 
  imuGravity: 9.79172227
  imuRPYWeight: 0.01
  extrinsicTrans: [0.049603512631594, -0.092737045313503, 0.015313004887488]
  extrinsicRot: [0.999788733066439, -0.011076602750357, 0.017314410031844,
                  0.011161790315359, 0.999926027879647, -0.004831549793495,
                  -0.017259609709402, 0.005023788111429, 0.999838422528471]
  extrinsicRPY: [0.999788733066439, -0.011076602750357, 0.017314410031844,
                  0.011161790315359, 0.999926027879647, -0.004831549793495,
                  -0.017259609709402, 0.005023788111429, 0.999838422528471]

  # LOAM feature threshold
  edgeThreshold: 1.0
  surfThreshold: 0.1
  edgeFeatureMinValidNum: 10
  surfFeatureMinValidNum: 100

  # voxel filter paprams
  odometrySurfLeafSize: 0.4                     # default: 0.4 - outdoor, 0.2 - indoor
  mappingCornerLeafSize: 0.2                    # default: 0.2 - outdoor, 0.1 - indoor
  mappingSurfLeafSize: 0.4                      # default: 0.4 - outdoor, 0.2 - indoor

  # robot motion constraint (in case you are using a 2D robot)
  z_tollerance: 1000                            # meters
  rotation_tollerance: 1000                     # radians

  # CPU Params
  numberOfCores: 4                              # number of cores for mapping optimization
  mappingProcessInterval: 0.15                  # seconds, regulate mapping frequency

  # Surrounding map
  surroundingkeyframeAddingDistThreshold: 1.0   # meters, regulate keyframe adding threshold
  surroundingkeyframeAddingAngleThreshold: 0.2  # radians, regulate keyframe adding threshold
  surroundingKeyframeDensity: 2.0               # meters, downsample surrounding keyframe poses   
  surroundingKeyframeSearchRadius: 50.0         # meters, within n meters scan-to-map optimization (when loop closure disabled)

  # Loop closure
  loopClosureEnableFlag: true
  loopClosureFrequency: 1.0                     # Hz, regulate loop closure constraint add frequency
  surroundingKeyframeSize: 50                   # submap size (when loop closure enabled)
  historyKeyframeSearchRadius: 15.0             # meters, key frame that is within n meters from current pose will be considerd for loop closure
  historyKeyframeSearchTimeDiff: 30.0           # seconds, key frame that is n seconds older will be considered for loop closure
  historyKeyframeSearchNum: 25                  # number of hostory key frames will be fused into a submap for loop closure
  historyKeyframeFitnessScore: 0.3              # icp threshold, the smaller the better alignment

  # Visualization
  globalMapVisualizationSearchRadius: 1000.0    # meters, global map visualization radius
  globalMapVisualizationPoseDensity: 10.0       # meters, global map visualization keyframe density
  globalMapVisualizationLeafSize: 1.0           # meters, global map visualization cloud density
